src/repositories/user.repository.ts

```diff
@@ -30,4 +41,41 @@ export class UserRepository extends DefaultCrudRepository<
       async () => orderRepository,
     );
   }
+
+  async login(credentials: Credentials) {
+    // Validate Email
+    if (!isemail.validate(credentials.email)) {
+      throw new HttpErrors.UnprocessableEntity('invalid email');
```

Comment:
Ah, I see there is a catch in what I proposed.

The Repository layer should not be coupled with HTTP/REST specifics. Instead of using `HttpErrors.UnprocessableEntity` and signal the particular error case via `statusCode` property, we should use just `err.code` here and have this code translated to transport specific means elsewhere. For example, gRPC may want to use a different mechanism to signal malformed request payload.

You can find the current translation table in `@loopback/rest` here:

https://github.com/strongloop/loopback-next/blob/e0f75ace756d8b5479ad5b66e9b2803e4b419eef/packages/rest/src/providers/reject.provider.ts#L12-L16
```ts
// TODO(bajtos) Make this mapping configurable at RestServer level,
// allow apps and extensions to contribute additional mappings.
const codeToStatusCodeMap: {[key: string]: number} = {
  ENTITY_NOT_FOUND: 404,
};
```

Unfortunately, we don't yet implemented API allowing extensions to contribute custom mappings. It's something we want to eventually provide, but there was no urgent need so far.

I am proposing the following approach here in this spike:
1. Make sure that all authorization-related errors come with an error code, e.g. `UNPROCESSABLE_ENTITY`.
2. For this spike, set `statusCode` too (by calling one of `HttpErrors` factories)
3. Add a comment to remind us that this solution is only temporary, and we need to make `codeToStatusCodeMap` customizable by extensions. Create a user story for this task, it will become a part of this spike's outcome.

See also https://github.com/strongloop/loopback-next/commit/7a56bad1430e7258bba137ce24d4fdaecdb394fd, https://github.com/strongloop/loopback-next/pull/1658 and https://github.com/strongloop/loopback-next/issues/1469

Reply:
@bajtos I found a really weird thing, the `Error` interface in `@types/node` doesn't have the `code` property, it only has an optional property `stack`. Which means by `let err = new Error('Wrong Credentials!')`, you cannot give `err` an error code. I could define a custom Error class that extend `Error` but if having to do that, I would rather leveraging the `HttpErrors` class from `rest` to create the error instance.

Next I am going to:

- Create a PR in `@types` to add the NodeJs Error properties like `code` in its `Error` interface.
- Create a new story to discuss 1. how to contribute a custom mapping in the reject action. 2. how to provide the mapping in a component. 3. update this example repo or the authentication module to add the mapping for the unauthorized error.

Reply:
> Create a PR in @types to add the NodeJs Error properties like code in its Error interface.

+1

Until that happens, you can use the following workaround:

```ts
const err = new HttpErrors.UnprocessableEntity('invalid email');
Object.assign(err, {code: 'INVALID_EMAIL'});
throw err;
```

> Create a new story to discuss 1. how to contribute a custom mapping in the reject action. 2. how to provide the mapping in a component. 3. update this example repo or the authentication module to add the mapping for the unauthorized error.

Awesome 👍

Reply:
This is a warning sign that `login` should be refactored out of the repo.